src/coreclr/src/inc/corprof.idl

```diff
@@ -4073,28 +4114,50 @@ interface ICorProfilerInfo11 : ICorProfilerInfo10
 ]
 interface ICorProfilerInfo12 : ICorProfilerInfo11
 {
+    HRESULT EventPipeStartSession(
+        [in]  UINT32                            cProviderConfigs,
+        [in, size_is(cProviderConfigs)]
+              COR_PRF_EVENTPIPE_PROVIDER_CONFIG pProviderConfigs[],
+        [in]  BOOL                              requestRundown,
+        [in]  BOOL                              requestCPUSamples,
```

Comment:
Looking at this logic, I think there probably is a few bugs here. Nobody every specified the enableSampleProfiler flag in EventPipe::Enable, it was a default arg with value of true. We were relying on SampleProfiler to check if the sampling event was enabled in SampleProfiler::Enable and bail early if it did.

The only place that ever checked for the presence of it was in the enable EventPipe via the environment variable path, but the existing delayed sampling initialization would only increase the SampleProfiler's ref count by 1 no matter how many sessions requested it before the runtime was far enough along.

For my refactoring I'm making EventPipe check if the sample provider is part of the provider list, and if it is unconditionally call SampleProvider::Enable. Then inside of SampleProvider it will delay sampling until the FinishInitialize is called, that way we can have an accurate refcount.